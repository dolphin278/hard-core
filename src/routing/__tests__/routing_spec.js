jest.unmock('..');

describe('Routing', function () {

  let Router;
  let router;

  beforeEach(() => {
    Router = require('..').Router;
    router = new Router();
  });

  describe('#add', () => {
    it('adding proper routes', () => {
      const handler = jest.fn();
      router.add('GET', '/about/', handler);
      expect(router.routes.length).toBe(1);
      expect(router.routes[0].method).toBe('GET');
      expect(router.routes[0].path).toBe('/about/');
      expect(router.routes[0].handler).toBe(handler);
    });

    it('adding proper routes without trailing slash', () => {
      const handler = jest.fn();
      router.add('GET', '/about', handler);
      expect(router.routes.length).toBe(1);
      expect(router.routes[0].method).toBe('GET');
      expect(router.routes[0].path).toBe('/about/');
      expect(router.routes[0].handler).toBe(handler);
    });

    it('adding proper routes without leading slash', () => {
      const handler = jest.fn();
      router.add('GET', 'about', handler);
      expect(router.routes.length).toBe(1);
      expect(router.routes[0].method).toBe('GET');
      expect(router.routes[0].path).toBe('/about/');
      expect(router.routes[0].handler).toBe(handler);
    });

    it('adding proper routes with lower case letters in method', () => {
      const handler = jest.fn();
      router.add('gEt', '/about', handler);
      expect(router.routes.length).toBe(1);
      expect(router.routes[0].method).toBe('GET');
      expect(router.routes[0].path).toBe('/about/');
      expect(router.routes[0].handler).toBe(handler);
    });

    it('adding unknown method must throw', () => {
      const handler = jest.fn();
      expect(function () {
        router.add('WRONG_METHOD', '/about', handler);
      }).toThrowError('Method WRONG_METHOD is not supported');
    });
  });

  describe('#buildTree', () => {
    it('builds tree', () => {
      var handler = function () {};
      router.add('GET', '/', handler);
      router.add('GET', '/about', handler);
      router.add('GET', '/posts', handler);
      router.add('GET', '/posts/:id', handler);
      router.add('GET', '/users/:id', handler);
      router.add('GET', '/users/', handler);
      router.add('GET', '/posts/:postId/comments/', handler);
      router.add('GET', '/posts/:postId/comments/:commentId', handler);
      router.add('GET', '/repos/:owner/:repo/keys/:id/', handler);
      router.add('GET', '/repos/:owner/:repo/issues/:number/labels/:name', handler);
      router.add('GET', '/my/super/long/about/page', handler);
      router.buildTree();
      expect(router.trees).toMatchSnapshot();
      // console.log(require('util').inspect(router.trees, {depth: 7, colors: true}))
    });
  });


  describe('#route', () => {
    let testTable;
    beforeEach(() => {
      testTable = [
        ['GET', '/', jest.fn(), [
          ['/', {}]
        ]],
        ['GET','/about/', jest.fn(), [
          ['/about', {}]
        ]],
        ['GET', '/posts', jest.fn(), [
          ['/posts', {}]
        ]],
        ['GET', '/posts/:id/', jest.fn(), [
          ['/posts/1', {id: '1'}],
          ['/posts/2/', {id: '2'}],
          ['/posts/abc/', {id: 'abc'}]
        ]],
        ['GET', '/posts/:postId/comments/', jest.fn(), [
          ['/posts/abc/comments', {postId: 'abc'}]
        ]],
        ['GET', '/posts/:postId/comments/:commentId/', jest.fn(), [
          ['/posts/abc/comments/1221', {postId: 'abc', commentId: '1221'}],
          ['/posts/abc/comments/1221/', {postId: 'abc', commentId: '1221'}],
          ['/posts/213/comments/122/', {postId: '213', commentId: '122'}],
        ]],
        ['GET', '/repos/:owner/:repo/keys/:id/', jest.fn(), [
          ['/repos/trekjs/trek/keys/233', {owner: 'trekjs', repo: 'trek', id: '233'}]
        ]],
        ['GET', '/repos/:owner/:repo/issues/:number/labels/:name/', jest.fn(), [
          ['/repos/trekjs/trek/issues/233/labels/help', {owner: 'trekjs', repo: 'trek', number:'233', name:'help'}]
        ]],
      ];
      testTable.reduce((a, [method, url, handler]) => router.add(method, url, handler), router);
      router.buildTree();
      // console.log(require('util').inspect(router.trees, {depth: 15, colors: true}))
    });

    it('test', () => {
      for (const [method, path, handler, tests] of testTable) {
        for (const [url, params] of tests) {
          const result = router.route(method, url);
          // console.log(method, url, require('util').inspect(result, {depth: 15, colors: true}));
          expect(typeof result).toBe('object');
          expect(result[0].route.handler).toBe(handler);
          expect(result[0].params).toEqual(params);
        }
      }
    });

  });

  describe('#compile', () => {
    let testTable;
    beforeEach(() => {
      testTable = [
        ['GET', '/', jest.fn(), [
          ['/', {}]
        ]],
        ['GET','/about/', jest.fn(), [
          ['/about', {}]
        ]],
        ['GET', '/posts', jest.fn(), [
          ['/posts', {}]
        ]],
        ['GET', '/posts/:id/', jest.fn(), [
          ['/posts/1', {id: '1'}],
          ['/posts/2/', {id: '2'}],
          ['/posts/abc/', {id: 'abc'}]
        ]],
        ['GET', '/posts/:postId/comments/', jest.fn(), [
          ['/posts/abc/comments', {postId: 'abc'}]
        ]],
        ['GET', '/posts/:postId/comments/:commentId/', jest.fn(), [
          ['/posts/abc/comments/1221', {postId: 'abc', commentId: '1221'}],
          ['/posts/abc/comments/1221/', {postId: 'abc', commentId: '1221'}],
          ['/posts/213/comments/122/', {postId: '213', commentId: '122'}],
        ]],
        ['GET', '/repos/:owner/:repo/keys/:id/', jest.fn(), [
          ['/repos/trekjs/trek/keys/233', {owner: 'trekjs', repo: 'trek', id: '233'}]
        ]],
        ['GET', '/repos/:owner/:repo/issues/:number/labels/:name/', jest.fn(), [
          ['/repos/trekjs/trek/issues/233/labels/help', {owner: 'trekjs', repo: 'trek', number:'233', name:'help'}]
        ]],
      ];
      testTable.reduce((a, [method, url, handler]) => router.add(method, url, handler), router);
      router.buildTree();
      // console.log(require('util').inspect(router.trees, {depth: 15, colors: true}))
    });
    it('compile', () => {
      router.compile();
    })

  });

  describe('#routeCompiled', () => {
    let testTable;
    beforeEach(() => {
      testTable = [
        ['GET', '/', jest.fn(), [
          ['/', {}]
        ]],
        ['GET','/about/', jest.fn(), [
          ['/about', {}]
        ]],
        ['GET', '/posts', jest.fn(), [
          ['/posts', {}]
        ]],
        ['GET', '/posts/:id/', jest.fn(), [
          ['/posts/1', {id: '1'}],
          ['/posts/2/', {id: '2'}],
          ['/posts/abc/', {id: 'abc'}]
        ]],
        ['GET', '/posts/:postId/comments/', jest.fn(), [
          ['/posts/abc/comments', {postId: 'abc'}]
        ]],
        ['GET', '/posts/:postId/comments/:commentId/', jest.fn(), [
          ['/posts/abc/comments/1221', {postId: 'abc', commentId: '1221'}],
          ['/posts/abc/comments/1221/', {postId: 'abc', commentId: '1221'}],
          ['/posts/213/comments/122/', {postId: '213', commentId: '122'}],
        ]],
        ['GET', '/repos/:owner/:repo/keys/:id/', jest.fn(), [
          ['/repos/trekjs/trek/keys/233', {owner: 'trekjs', repo: 'trek', id: '233'}]
        ]],
        ['GET', '/repos/:owner/:repo/issues/:number/labels/:name/', jest.fn(), [
          ['/repos/trekjs/trek/issues/233/labels/help', {owner: 'trekjs', repo: 'trek', number:'233', name:'help'}]
        ]],
      ];
      testTable.reduce((a, [method, url, handler]) => router.add(method, url, handler), router);
      router.buildTree();
      router.compile();
      // console.loga(require('util').inspect(router.trees, {depth: 15, colors: true}))
    });

    it('test', () => {
      for (const [method, _, handler, tests] of testTable) {
        for (const [url, params] of tests) {
          const result = router.routeCompiled(method, url);
          // console.log(method, url, require('util').inspect(result, {depth: 15, colors: true}));
          expect(typeof result).toBe('object');
          expect(result[0].route.handler).toBe(handler);
          expect(result[0].params).toEqual(params);
        }
      }
    });

  });

});
