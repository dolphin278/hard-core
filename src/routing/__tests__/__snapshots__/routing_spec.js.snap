// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routing #buildTree builds tree 1`] = `
Object {
  "GET": Node {
    "children": Object {
      "about": Node {
        "children": Object {},
        "paramNames": Object {},
        "routes": Array [
          Route {
            "handler": [Function],
            "id": 1,
            "method": "GET",
            "parametersMap": Array [],
            "path": "/about/",
          },
        ],
        "value": "about",
      },
      "my": Node {
        "children": Object {
          "super": Node {
            "children": Object {
              "long": Node {
                "children": Object {
                  "about": Node {
                    "children": Object {
                      "page": Node {
                        "children": Object {},
                        "paramNames": Object {},
                        "routes": Array [
                          Route {
                            "handler": [Function],
                            "id": 10,
                            "method": "GET",
                            "parametersMap": Array [],
                            "path": "/my/super/long/about/page/",
                          },
                        ],
                        "value": "page",
                      },
                    },
                    "paramNames": Object {},
                    "routes": Array [],
                    "value": "about",
                  },
                },
                "paramNames": Object {},
                "routes": Array [],
                "value": "long",
              },
            },
            "paramNames": Object {},
            "routes": Array [],
            "value": "super",
          },
        },
        "paramNames": Object {},
        "routes": Array [],
        "value": "my",
      },
      "posts": Node {
        "children": Object {
          "*": Node {
            "children": Object {
              "comments": Node {
                "children": Object {
                  "*": Node {
                    "children": Object {},
                    "paramNames": Object {},
                    "routes": Array [
                      Route {
                        "handler": [Function],
                        "id": 7,
                        "method": "GET",
                        "parametersMap": Array [
                          Array [
                            "postId",
                            0,
                          ],
                          Array [
                            "commentId",
                            1,
                          ],
                        ],
                        "path": "/posts/:postId/comments/:commentId/",
                      },
                    ],
                    "value": "*",
                  },
                },
                "paramNames": Object {
                  "7": "commentId",
                },
                "routes": Array [
                  Route {
                    "handler": [Function],
                    "id": 6,
                    "method": "GET",
                    "parametersMap": Array [
                      Array [
                        "postId",
                        0,
                      ],
                    ],
                    "path": "/posts/:postId/comments/",
                  },
                ],
                "value": "comments",
              },
            },
            "paramNames": Object {},
            "routes": Array [
              Route {
                "handler": [Function],
                "id": 3,
                "method": "GET",
                "parametersMap": Array [
                  Array [
                    "id",
                    0,
                  ],
                ],
                "path": "/posts/:id/",
              },
            ],
            "value": "*",
          },
        },
        "paramNames": Object {
          "3": "id",
          "6": "postId",
          "7": "postId",
        },
        "routes": Array [
          Route {
            "handler": [Function],
            "id": 2,
            "method": "GET",
            "parametersMap": Array [],
            "path": "/posts/",
          },
        ],
        "value": "posts",
      },
      "repos": Node {
        "children": Object {
          "*": Node {
            "children": Object {
              "*": Node {
                "children": Object {
                  "issues": Node {
                    "children": Object {
                      "*": Node {
                        "children": Object {
                          "labels": Node {
                            "children": Object {
                              "*": Node {
                                "children": Object {},
                                "paramNames": Object {},
                                "routes": Array [
                                  Route {
                                    "handler": [Function],
                                    "id": 9,
                                    "method": "GET",
                                    "parametersMap": Array [
                                      Array [
                                        "owner",
                                        0,
                                      ],
                                      Array [
                                        "repo",
                                        1,
                                      ],
                                      Array [
                                        "number",
                                        2,
                                      ],
                                      Array [
                                        "name",
                                        3,
                                      ],
                                    ],
                                    "path": "/repos/:owner/:repo/issues/:number/labels/:name/",
                                  },
                                ],
                                "value": "*",
                              },
                            },
                            "paramNames": Object {
                              "9": "name",
                            },
                            "routes": Array [],
                            "value": "labels",
                          },
                        },
                        "paramNames": Object {},
                        "routes": Array [],
                        "value": "*",
                      },
                    },
                    "paramNames": Object {
                      "9": "number",
                    },
                    "routes": Array [],
                    "value": "issues",
                  },
                  "keys": Node {
                    "children": Object {
                      "*": Node {
                        "children": Object {},
                        "paramNames": Object {},
                        "routes": Array [
                          Route {
                            "handler": [Function],
                            "id": 8,
                            "method": "GET",
                            "parametersMap": Array [
                              Array [
                                "owner",
                                0,
                              ],
                              Array [
                                "repo",
                                1,
                              ],
                              Array [
                                "id",
                                2,
                              ],
                            ],
                            "path": "/repos/:owner/:repo/keys/:id/",
                          },
                        ],
                        "value": "*",
                      },
                    },
                    "paramNames": Object {
                      "8": "id",
                    },
                    "routes": Array [],
                    "value": "keys",
                  },
                },
                "paramNames": Object {},
                "routes": Array [],
                "value": "*",
              },
            },
            "paramNames": Object {
              "8": "repo",
              "9": "repo",
            },
            "routes": Array [],
            "value": "*",
          },
        },
        "paramNames": Object {
          "8": "owner",
          "9": "owner",
        },
        "routes": Array [],
        "value": "repos",
      },
      "users": Node {
        "children": Object {
          "*": Node {
            "children": Object {},
            "paramNames": Object {},
            "routes": Array [
              Route {
                "handler": [Function],
                "id": 4,
                "method": "GET",
                "parametersMap": Array [
                  Array [
                    "id",
                    0,
                  ],
                ],
                "path": "/users/:id/",
              },
            ],
            "value": "*",
          },
        },
        "paramNames": Object {
          "4": "id",
        },
        "routes": Array [
          Route {
            "handler": [Function],
            "id": 5,
            "method": "GET",
            "parametersMap": Array [],
            "path": "/users/",
          },
        ],
        "value": "users",
      },
    },
    "paramNames": Object {},
    "routes": Array [
      Route {
        "handler": [Function],
        "id": 0,
        "method": "GET",
        "parametersMap": Array [],
        "path": "/",
      },
    ],
    "value": "",
  },
}
`;
